File: epi_judge_java/epi/ThreeSum.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.Collections;
import java.util.List;

public class ThreeSum {

    @EpiTest(testDataFile = "three_sum.tsv")
    public static boolean hasThreeSum(List<Integer> nums, int t) {
        nums.sort(null);
        for (int i = 0; i < nums.size(); i++) {
            for (int j = i; j < nums.size(); j++) {
                if (Collections.binarySearch(nums, t - nums.get(i) - nums.get(j)) >= 0) return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "ThreeSum.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/three_sum.cc
#include <vector>

#include "test_framework/generic_test.h"

using namespace std;

bool HasThreeSum(vector<int> nums, int t) {
    sort(nums.begin(), nums.end());
    for (int start = 0; start < nums.size(); start++) {
        int left = start;
        int right = nums.size() - 1;
        while (left <= right) {
            int sum = nums[start] + nums[left] + nums[right];
            if (sum > t) {
                right--;
            } else if (sum < t) {
                left++;
            } else {
                return  true;
            }
        }
    }

    return false;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"A", "t"};
    return GenericTestMain(args, "three_sum.cc", "three_sum.tsv", &HasThreeSum,
                           DefaultComparator{}, param_names);
}
